allennlp/tests/training/metrics/spearman_correlation_test.py

```diff
@@ -0,0 +1,123 @@
+# coding=utf-8
+# @Author: wb-gmh583075
+# @Date: 2019-07-31
+
+import torch
+import numpy as np
+from numpy.testing import assert_allclose
+
+from allennlp.common.testing import AllenNlpTestCase
+from allennlp.training.metrics import SpearmanCorrelation
+
+
+def spearman_corrcoef(predictions, labels, mask=None):
+    if mask is not None:
+        predictions = predictions * mask
+        labels = labels * mask
+
+    # if all number of a set is same, return np.nan
+    if len(np.unique(predictions)) == 1 or len(np.unique(labels)) == 1:
+        return np.nan
+
+    n = len(predictions)
+
+    predictions = [(k, v) for k, v in enumerate(predictions)]
+    predictions.sort(key=lambda x: x[1], reverse=True)
+    predictions = [(k, v) for k, v in enumerate(predictions)]
+    predictions.sort(key=lambda x: x[1][0])
+
+    labels = [(k, v) for k, v in enumerate(labels)]
+    labels.sort(key=lambda x: x[1], reverse=True)
+    labels = [(k, v) for k, v in enumerate(labels)]
+    labels.sort(key=lambda x: x[1][0])
+
+    total = 0
+    for i in range(n):
+        total += (predictions[i][0] - labels[i][0]) ** 2
+    expected_spearman_correlation = 1 - float(6 * total) / (n * (n ** 2 - 1))
+
+    return expected_spearman_correlation
+
+
+class SpearmanCorrelationTest(AllenNlpTestCase):
+    def test_spearman_correlation_unmasked_computation(self):
+        spearman_correlation = SpearmanCorrelation()
+        batch_size = 100
+        num_labels = 10
+        predictions1 = np.random.randn(batch_size, num_labels).astype("float32")
+        labels1 = 0.5 * predictions1 + np.random.randn(batch_size, num_labels).astype("float32")
+
+        predictions2 = np.random.randn(1).repeat(num_labels).astype("float32")
+        predictions2 = predictions2[np.newaxis, :].repeat(batch_size, axis=0)
+        labels2 = np.random.randn(1).repeat(num_labels).astype("float32")
+        labels2 = 0.5 * predictions2 + labels2[np.newaxis, :].repeat(batch_size, axis=0)
+
+        # in top 100 batch, 10 number in each batch is different; The last 100 batch, 10 number in each batch is same.
+        predictions_labels_ = [(predictions1, labels1), (predictions2, labels2)]
+
+        stride = 10

```

Comment:
Generally if these tests could be simpler (no striding, just test the case with and without masks for random tensors). In particular, the tests do not call the metric with _batched_ tensors, which would be good (rather than having the doubly nested for loops).

Reply:
ok